refactor(logging): enhance log messages for device resource counting

haitwang-cloud opened the pull request:
**What type of PR is this?**

/kind flake

**What this PR does / why we need it**:

The current logging configuration of our GPU allocation system is not sufficient to track device resource allocation and request processing, making it challenging to debug and troubleshoot issues. This PR aims to enhance the logging functionality to provide a more comprehensive understanding of device resource utilization and request processing.

The current log output for a request processing event is as follows:

```
I1223 05:51:46.777241       1 route.go:44] Into Predicate Route inner func
I1223 05:51:46.777693       1 scheduler.go:437] "begin schedule filter" pod="cuda-sample-vector-add-78dd7b6c56-jkxtp" uuid="064360f8-b1d5-465c-a23b-675832a43899" namespaces="gpu-test-workloads"
I1223 05:51:46.777728       1 device.go:213] Counting mthreads devices
I1223 05:51:46.777736       1 device.go:123] Counting metax devices
I1223 05:51:46.777744       1 device.go:254] Counting mlu devices
I1223 05:51:46.777749       1 device.go:259] idx= nvidia.com/gpu val= {{1 0} {<nil>} 1 DecimalSI} {{0 0} {<nil>}  }
I1223 05:51:46.777761       1 device.go:259] idx= nvidia.com/gpumem val= {{4 3} {<nil>} 4k DecimalSI} {{0 0} {<nil>}  }
I1223 05:51:46.777767       1 device.go:188] Counting dcu devices
I1223 05:51:46.777771       1 device.go:183] Counting iluvatar devices
I1223 05:51:46.777788       1 pod.go:40] "collect requestreqs" counts=[{"NVIDIA":{"Nums":1,"Type":"NVIDIA","Memreq":4000,"MemPercentagereq":101,"Coresreq":25}}]
I1223 05:51:46.777823       1 scheduler.go:471] All node scores do not meet for pod cuda-sample-vector-add-78dd7b6c56-jkxtp ]
```

However, this log output lacks critical information about device resource allocation and request processing, making it difficult to diagnose issues. This PR introduces detailed logs for device resource counting and request processing, as well as improved error handling and logging mechanisms.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
Please review the updated logging configuration and ensure that it meets the desired logging requirements.
**Does this PR introduce a user-facing change?**:
No

haitwang-cloud commented:
/kind flake

haitwang-cloud commented:
Use this PR to replace the old one https://github.com/Project-HAMi/HAMi/pull/732 ( with conflicting with master branch issue)

wawa0210 commented:
There are many changes, and the UT coverage rate has decreased, which we need to pay attention to.

haitwang-cloud commented:
Appending vGPU alloc result
```
(base) root@tim-demo--001-0:~# date
Tue Jan  7 06:13:40 AM UTC 2025
(base) root@tim-demo--001-0:~# nvidia-smi
[HAMI-core Msg(154:140251316950848:libvgpu.c:836)]: Initializing.....
Tue Jan  7 06:13:44 2025       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.86.10              Driver Version: 535.86.10    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla V100-PCIE-16GB           Off | 00000000:D8:00.0 Off |                    0 |
| N/A   31C    P0              27W / 250W |      0MiB /  1638MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
[HAMI-core Msg(154:140251316950848:multiprocess_memory_limit.c:497)]: Calling exit handler 154
```

haitwang-cloud commented:
Append the scheduler log

```
Starting to collect metrics for scheduler
I0107 06:16:07.160303       1 pods.go:107] Getting all scheduled pods with 7 nums
I0107 06:16:07.160322       1 metrics.go:192] Collecting dask shoot--ait--datalake-nonprod-c32m384-v100g3-z1-977d9-ctcpp worker-6cc67f6f54-zb9jw GPU-a346704d-b53d-8ef3-eb65-113a9d671130 25 8000
I0107 06:16:07.160351       1 metrics.go:192] Collecting ism shoot--ait--datalake-nonprod-c32m384-v100g3-z1-977d9-ctcpp deploy-ism-embedding-578cd4ff95-cc9bn GPU-59f8a413-1c74-960f-18e3-d32f52626071 25 3000
I0107 06:16:07.160377       1 metrics.go:192] Collecting ism shoot--ait--datalake-nonprod-c32m384-v100g3-z1-977d9-ctcpp dev-deploy-ism-embedding-5794697b4f-7xnx9 GPU-a346704d-b53d-8ef3-eb65-113a9d671130 25 3000
I0107 06:16:07.160404       1 metrics.go:192] Collecting kbf-i064931 shoot--ait--datalake-nonprod-c32m384-v100g3-z1-977d9-ctcpp nirjar-lrs-gpu-0 GPU-c5d16691-bb12-0aad-bb52-9238619d1ad6 25 3000
I0107 06:16:07.160435       1 metrics.go:192] Collecting kbf-i744932 shoot--ait--datalake-nonprod-c32m384-v100g3-z1-977d9-ctcpp pyspark-img-v12-0 GPU-59f8a413-1c74-960f-18e3-d32f52626071 25 3000
I0107 06:16:07.160465       1 metrics.go:192] Collecting sir-service shoot--ait--datalake-nonprod-c32m384-v100g3-z1-977d9-ctcpp sir-core-service-5f94bdfbb-xcbcd GPU-c5d16691-bb12-0aad-bb52-9238619d1ad6 25 3000
I0107 06:16:07.160490       1 metrics.go:192] Collecting kbf-i742968 shoot--ait--datalake-nonprod-c32m384-v100g3-z1-977d9-ctcpp tim-demo--001-0 GPU-c5d16691-bb12-0aad-bb52-9238619d1ad6 25 1638
```

haitwang-cloud commented:
Append the hami-device-plugin log

```
I0107 06:21:41.194912 2733973 main.go:298] Retrieving plugins.
I0107 06:21:41.196638 2733973 factory.go:123] Detected NVML platform: found NVML library
I0107 06:21:41.196741 2733973 factory.go:123] Detected non-Tegra platform: /sys/devices/soc0/family file not found
I0107 06:21:41.230607 2733973 server.go:145] reading config= &{0xc00061e0b0 0xc000253c80 <nil>} resourceName 0xc000253c80 configfile= /device-config.yaml sconfig= &{{nvidia.com/gpu nvidia.com/gpumem nvidia.com/gpucores nvidia.com/gpumem-percentage nvidia.com/priority false 3000 25 1 10 1 1 false []} {metax-tech.com/gpu} {hygon.com/dcunum hygon.com/dcumem hygon.com/dcucores} {cambricon.com/vmlu cambricon.com/mlu.smlu.vmemory cambricon.com/mlu.smlu.vcore} {mthreads.com/vgpu mthreads.com/sgpu-memory mthreads.com/sgpu-core} {iluvatar.ai/vgpu iluvatar.ai/vcuda-memory iluvatar.ai/vcuda-core} [{Ascend910A 910B huawei.com/Ascend910A huawei.com/Ascend910A-memory 32768 32768 30 0 [{vir02 2184 2 0} {vir04 4369 4 0} {vir08 8738 8 0} {vir16 17476 16 0}]} {Ascend910B 910B3 huawei.com/Ascend910B huawei.com/Ascend910B-memory 65536 65536 20 7 [{vir05_1c_16g 16384 5 1} {vir10_3c_32g 32768 10 3}]} {Ascend310P 310P3 huawei.com/Ascend310P huawei.com/Ascend310P-memory 21527 24576 8 7 [{vir01 3072 1 1} {vir02 6144 2 2} {vir04 12288 4 4}]}]}
I0107 06:21:41.230928 2733973 server.go:113] Device Plugin Configs: {[{m5-cloudinfra-online02  1.8 0 10 none 0xc000190300}]}
I0107 06:21:41.230992 2733973 device.go:128] "initializing nvidia device" resourceName="nvidia.com/gpu" resourceMem="nvidia.com/gpumem" DefaultGPUNum=1
I0107 06:21:41.231492 2733973 server.go:278] Starting GRPC server for 'nvidia.com/gpu'
I0107 06:21:41.232766 2733973 server.go:204] Starting to serve 'nvidia.com/gpu' on /var/lib/kubelet/device-plugins/nvidia-gpu.sock
I0107 06:21:41.239438 2733973 server.go:212] Registered device plugin for 'nvidia.com/gpu' with Kubelet
I0107 06:21:41.239508 2733973 register.go:200] Starting WatchAndRegister
I0107 06:21:41.257706 2733973 register.go:144] MemoryScaling= 1 registeredmem= 16384
I0107 06:21:41.384372 2733973 register.go:173] nvml registered device id=1, memory=16384, type=Tesla V100-PCIE-16GB, numa=0
I0107 06:21:41.384734 2733973 register.go:144] MemoryScaling= 1 registeredmem= 16384
I0107 06:21:41.509410 2733973 register.go:173] nvml registered device id=2, memory=16384, type=Tesla V100-PCIE-16GB, numa=0
I0107 06:21:41.509558 2733973 register.go:144] MemoryScaling= 1 registeredmem= 16384
I0107 06:21:41.634646 2733973 register.go:173] nvml registered device id=0, memory=16384, type=Tesla V100-PCIE-16GB, numa=0
I0107 06:21:41.634936 2733973 register.go:180] "start working on the devices" devices=[{"id":"GPU-59f8a413-1c74-960f-18e3-d32f52626071","index":1,"count":10,"devmem":16384,"devcore":100,"type":"NVIDIA-Tesla V100-PCIE-16GB","mode":"hami-core","health":true},{"id":"GPU-c5d16691-bb12-0aad-bb52-9238619d1ad6","index":2,"count":10,"devmem":16384,"devcore":100,"type":"NVIDIA-Tesla V100-PCIE-16GB","mode":"hami-core","health":true},{"id":"GPU-a346704d-b53d-8ef3-eb65-113a9d671130","count":10,"devmem":16384,"devcore":100,"type":"NVIDIA-Tesla V100-PCIE-16GB","mode":"hami-core","health":true}]
I0107 06:21:41.674585 2733973 register.go:190] patch node with the following annos map[hami.io/node-handshake:Reported 2025-01-07 06:21:41.674551147 +0000 UTC m=+0.503881448 hami.io/node-nvidia-register:GPU-59f8a413-1c74-960f-18e3-d32f52626071,10,16384,100,NVIDIA-Tesla V100-PCIE-16GB,0,true,1,hami-core:GPU-c5d16691-bb12-0aad-bb52-9238619d1ad6,10,16384,100,NVIDIA-Tesla V100-PCIE-16GB,0,true,2,hami-core:GPU-a346704d-b53d-8ef3-eb65-113a9d671130,10,16384,100,NVIDIA-Tesla V100-PCIE-16GB,0,true,0,hami-core:]
I0107 06:21:41.697963 2733973 register.go:210] Successfully registered annotation. Next check in 30s seconds...
I0107 06:22:11.698908 2733973 register.go:144] MemoryScaling= 1 registeredmem= 16384
I0107 06:22:11.826514 2733973 register.go:173] nvml registered device id=0, memory=16384, type=Tesla V100-PCIE-16GB, numa=0
I0107 06:22:11.826705 2733973 register.go:144] MemoryScaling= 1 registeredmem= 16384
I0107 06:22:11.948941 2733973 register.go:173] nvml registered device id=1, memory=16384, type=Tesla V100-PCIE-16GB, numa=0
I0107 06:22:11.949196 2733973 register.go:144] MemoryScaling= 1 registeredmem= 16384
I0107 06:22:12.072465 2733973 register.go:173] nvml registered device id=2, memory=16384, type=Tesla V100-PCIE-16GB, numa=0
I0107 06:22:12.072555 2733973 register.go:180] "start working on the devices" devices=[{"id":"GPU-a346704d-b53d-8ef3-eb65-113a9d671130","count":10,"devmem":16384,"devcore":100,"type":"NVIDIA-Tesla V100-PCIE-16GB","mode":"hami-core","health":true},{"id":"GPU-59f8a413-1c74-960f-18e3-d32f52626071","index":1,"count":10,"devmem":16384,"devcore":100,"type":"NVIDIA-Tesla V100-PCIE-16GB","mode":"hami-core","health":true},{"id":"GPU-c5d16691-bb12-0aad-bb52-9238619d1ad6","index":2,"count":10,"devmem":16384,"devcore":100,"type":"NVIDIA-Tesla V100-PCIE-16GB","mode":"hami-core","health":true}]
I0107 06:22:12.088944 2733973 register.go:190] patch node with the following annos map[hami.io/node-handshake:Reported 2025-01-07 06:22:12.088894476 +0000 UTC m=+30.918224765 hami.io/node-nvidia-register:GPU-a346704d-b53d-8ef3-eb65-113a9d671130,10,16384,100,NVIDIA-Tesla V100-PCIE-16GB,0,true,0,hami-core:GPU-59f8a413-1c74-960f-18e3-d32f52626071,10,16384,100,NVIDIA-Tesla V100-PCIE-16GB,0,true,1,hami-core:GPU-c5d16691-bb12-0aad-bb52-9238619d1ad6,10,16384,100,NVIDIA-Tesla V100-PCIE-16GB,0,true,2,hami-core:]
I0107 06:22:12.110247 2733973 register.go:210] Successfully registered annotation. Next check in 30s seconds...
```

wawa0210 commented on .DS_Store:
Was this file included by mistake?

haitwang-cloud replied:
Oops, removed this file

haitwang-cloud commented:
This pull request includes several changes across multiple files to improve error handling, refactor code, and enhance testing. The main changes involve adding error handling for device initialization, refactoring the `GenerateMigTemplate` function, and updating test cases for better coverage.

### Error Handling Improvements:
* [`pkg/device-plugin/nvidiadevice/nvinternal/plugin/server.go`](diffhunk://#diff-4beccb94da8e027b37416f87ed5a76a77a5ea6ea62c9a69a8b1833816754fe5aL153-R156): Added error handling for device initialization in `NewNvidiaDevicePlugin` to log a fatal error if initialization fails.

### Code Refactoring:
* [`pkg/device-plugin/nvidiadevice/nvinternal/plugin/util.go`](diffhunk://#diff-8b63c6d64d45d3fd4890aafbcf25428f0747080081d88dc8d6d73cee29e31dd5L208-R269): Refactored the `GenerateMigTemplate` function to improve readability and added helper functions `containsModel` and `containsDevice` for better code organization.

### Test Enhancements:
* [`pkg/device-plugin/nvidiadevice/nvinternal/plugin/util_test.go`](diffhunk://#diff-b468dce79224c99922450b4bb0379e42c17927fb7552ba1c2165aac8732c9b5dL149-R155): Updated the `TestGenerateMigTemplate` function to use a table-driven approach for better test coverage and clarity.

### Import Cleanup:
* [`pkg/device-plugin/nvidiadevice/nvinternal/plugin/util.go`](diffhunk://#diff-8b63c6d64d45d3fd4890aafbcf25428f0747080081d88dc8d6d73cee29e31dd5L33): Removed an unused import statement.
* [`pkg/device/cambricon/device_test.go`](diffhunk://#diff-527a4b05c0342ab49b990245bab37060848a59b6a17ac7dea9269f90b879bf42L25-R32): Reorganized import statements for better readability.

### Logging Improvements:
* [`pkg/device/cambricon/device.go`](diffhunk://#diff-18913db9fdb46323a9ebfa1c20bc35d6d05a0ab7948f00e1a2b4d1e953f90bc2L253-R253): Improved logging messages in the `GenerateResourceRequests` function to include the container name.
